### PR TITLE
feat: `ProviderFactory`

### PR DESCRIPTION
**Pull request by detectivekim**
Implement `ProviderFactory`.

### CODE REVIEW

File: src/Provider.sol [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (JhChoy) [inline src/Provider.sol]: It'll be better to change the license to GPL.

File: src/Provider.sol [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import {IProviderFactory} from "./interfaces/IProviderFactory.sol";
+
+contract Provider is IProvider {
+    using CurrencyLibrary for Currency;
+
+    uint256 public constant RATE_PRECISION = 10 ** 6;
+
+    IBookManager private immutable _bookManager;
+    IProviderFactory public immutable factory;
+    address public immutable broker;
```
Comment (JhChoy) [inline src/Provider.sol:17]: How about `recipient`?

File: src/Provider.sol [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    uint256 public immutable shareRatio;
+
+    constructor(address broker_, uint256 shareRatio_) {
+        factory = IProviderFactory(msg.sender);
+        _bookManager = factory.bookManager();
+        broker = broker_;
+        shareRatio = shareRatio_;
+    }
+
+    function claim(Currency[] calldata currencies) external {
```
Comment (JhChoy) [inline src/Provider.sol:27]: I think we should add `ReentrancyGuard` here.

File: src/Provider.sol [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        address protocolTreasury = factory.treasury();
+        for (uint256 i = 0; i < currencies.length; i++) {
```
Comment (JhChoy) [inline src/Provider.sol]:
```suggestion
        for (uint256 i = 0; i < currencies.length; ++i) {
```

File: src/interfaces/IProvider.sol [hunk 5]
```diff
@@ -0,0 +1,17 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import {Currency, CurrencyLibrary} from "v2-core/libraries/Currency.sol";
+import {IProviderFactory} from "./IProviderFactory.sol";
+
+interface IProvider {
+    event Claim(address indexed broker, address indexed protocolTreasury, uint256 brokerShare, uint256 protocolShare);
```
Comment (JhChoy) [inline src/interfaces/IProvider.sol]: I believe `currency` should be added here too.

File: src/ProviderFactory.sol [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        bookManager.whitelist(provider);
+        emit DeployProvider(provider, broker, shareRatio);
+    }
+
+    function setTreasury(address newTreasury) external onlyOwner {
+        treasury = newTreasury;
+        emit SetTreasury(newTreasury);
+    }
+
+    /**
```
Comment (JhChoy) [inline src/ProviderFactory.sol]: Theses comments should be moved to interface.

File: test/unit/Provider.t.sol [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+contract ProviderTest is Test {
+    ProviderFactory public providerFactory;
+    address public bookManager;
+
+    function setUp() public {
+        bookManager = address(new MockBookManager(address(this)));
+        vm.deal(bookManager, 10 ether);
+        providerFactory = new ProviderFactory();
+        Ownable2Step(bookManager).transferOwnership(address(providerFactory));
+        providerFactory.__ProviderFactory_init(address(this), address(bookManager), Constants.TREASURY, 300000);
```
Comment (JhChoy) [inline test/unit/Provider.t.sol]: Testing with the proxy looks much better.

**Review by JhChoy — COMMENTED** [review]
Also, some tests for the `ProviderFactory` are needed.

File: src/interfaces/IProvider.sol (hunk 5 above)
Comment (detectivekim) [reply src/interfaces/IProvider.sol]: Resolved.

File: src/ProviderFactory.sol (hunk 6 above)
Comment (detectivekim) [reply src/ProviderFactory.sol]: Resolved.

File: src/Provider.sol (hunk 1 above)
Comment (detectivekim) [reply src/Provider.sol]: Resolved.

**Review by JhChoy — APPROVED** [review]

**Pull request merged**